client/app/components/QuerySelector.jsx

```diff
@@ -0,0 +1,160 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import { debounce, find } from 'lodash';
+import Input from 'antd/lib/input';
+import { Query } from '@/services/query';
+import { toastr } from '@/services/ng';
+import { QueryTagsControl } from '@/components/tags-control/TagsControl';
+
+const SEARCH_DEBOUNCE_DURATION = 200;
+
+class StaleSearchError extends Error {
+  constructor() {
+    super('stale search');
+  }
+}
+
+function search(term) {
+  // get recent
+  if (!term) {
+    return Query.recent().$promise
+      .then((results) => {
+        const filteredResults = results.filter(item => !item.is_draft); // filter out draft
+        return Promise.resolve(filteredResults);
+      });
+  }
+
+  // search by query
+  return Query.query({ q: term }).$promise
+    .then(({ results }) => Promise.resolve(results));
+}
+
+export function QuerySelector(props) {
+  const [searchTerm, setSearchTerm] = useState();
+  const [searching, setSearching] = useState();
+  const [searchResults, setSearchResults] = useState([]);
+  const [selectedQuery, setSelectedQuery] = useState();
+
+  let isStaleSearch = false;
+  const debouncedSearch = debounce(_search, SEARCH_DEBOUNCE_DURATION);
+  const placeholder = 'Search a query by name';
+  const clearIcon = <i className="fa fa-times" onClick={() => selectQuery(null)} />;
+  const spinIcon = <i className={cx('fa fa-spinner fa-pulse', { hidden: !searching })} />;
+
+  // set selected from prop
+  useEffect(() => {
+    if (props.selectedQuery) {
+      setSelectedQuery(props.selectedQuery);
+    }
+  }, [props.selectedQuery]);
+
+  // on search term changed, debounced
+  useEffect(() => {
+    // clear results, no search
+    if (searchTerm === null) {
+      setSearchResults(null);
+      return () => {};
+    }
+
+    // search
+    debouncedSearch(searchTerm);
+    return () => {
+      debouncedSearch.cancel();
+      isStaleSearch = true;
+    };
+  }, [searchTerm]);
```

Comment:
@ranbena Look: you still load search results as a side-effect of changing search term. It's implicit dependency, that's what makes code hard to read and understand:
```javascript
$scope.setSearchTerm = (searchTerm) => {
  $scope.searchTerm = searchTerm;
};

// somewhere else

$scope.$watch('searchTerm', (searchTerm) => {
  if (searchTerm == '') {
    $scope.results = [];
  } else {
    doSearch(searchTerm);
  }
});
```
in opposite to explicit reaction, where you immediately see that if you change search term - it will do some extra stuff:
```javascript
$scope.setSearchTerm = (searchTerm) => {
  $scope.searchTerm = searchTerm;
  if (searchTerm == '') {
    $scope.results = [];
  } else {
    doSearch(searchTerm);
  }
};
```
I intentionally provide examples in Angular because this pattern is really common there, and one of best practices in React for a long time was to avoid such code.

Reply:
Since this useEffect also allows easy stale/debounce management (the alternative I assume is having dedicated refs), I'd rather keep it and keep moving 🏃